File: frontend/utils/validation.ts
export const PAN_RE = /^[A-Z]{5}[0-9]{4}[A-Z]{1}$/;
export const AADHAAR_RE = /^\d{12}$/;
export const EMAIL_RE = /^[^@\s]+@[^@\s]+\.[^@\s]+$/;
export const OTP_RE = /^\d{4,6}$/;

export const validate = (name: string, value: string): string | null => {
  const n = name.toLowerCase();
  if (n.includes("pan") && !PAN_RE.test(value)) return "Invalid PAN format (e.g., ABCDE1234F)";
  if ((n.includes("aadhaar") || n.includes("aadhar")) && !AADHAAR_RE.test(value)) return "Aadhaar must be 12 digits";
  if (n.includes("email") && !EMAIL_RE.test(value)) return "Invalid email address";
  if (n.includes("otp") && !OTP_RE.test(value)) return "OTP must be 4–6 digits";
  return null;
};

File: backend/src/validation.py
import re


OTP_STORE = {}

PAN_RE = re.compile(r"^[A-Z]{5}[0-9]{4}[A-Z]{1}$")
AADHAAR_RE = re.compile(r"^\d{12}$")
EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")
OTP_RE = re.compile(r"^\d{4,6}$")

def validate_pan(pan: str) -> bool:
    return bool(PAN_RE.match(pan or ""))

def validate_aadhaar(aadhaar: str) -> bool:
    return bool(AADHAAR_RE.match(aadhaar or ""))

def validate_email(email: str | None) -> bool:
    if not email:
        return True
    return bool(EMAIL_RE.match(email))
    
def validate_otp(otp: str | None) -> bool:
    if not otp:
        return True
    return bool(OTP_RE.match(otp))
